src/components/Settings/Context.tsx

```diff
@@ -0,0 +1,254 @@
+import {
+  Context,
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useReducer,
+} from "react";
+import debounce from "lodash.debounce";
+import isEqual from "lodash.isequal";
+import { SpruceFormProps } from "components/SpruceForm/types";
+import { FormToGqlFunction, SettingsRoutes } from "./types";
+
+type OnChangeParams<
+  T extends SettingsRoutes,
+  FormStateMap extends Record<T, any>
+> = Pick<
+  Parameters<SpruceFormProps<FormStateMap[T]>["onChange"]>[0],
+  "formData" | "errors"
+>;
+
+// TypeScript has a bug preventing the formData type mapping from working correctly.
+// https://github.com/microsoft/TypeScript/issues/24085
+// For now, leave as-is and assert form state types when errors are thrown.
+export type TabState<
+  T extends SettingsRoutes,
+  FormStateMap extends Record<T, any>
```

Comment:
Not effectively, imo. RJSF never types `formData` more strictly than `any` ([here](https://github.com/rjsf-team/react-jsonschema-form/blob/v4.2.3/packages/core/index.d.ts)). Also, we do end up with more specific types for every `formData` prop since this `any` just used for `extends`.